### ISSUE TITLE
CSV export is not aggregating payroll hours for the full range of dates included in a payroll run.

### ISSUE THREAD

**Issue by Mei152** [Mei152/hours_import#123]

**Comment by Mei152** [comment]
Something is up with the way postgres is interpreting the end date in the aggregation query. 
`time_records.clockin BETWEEN '2016-03-21' AND '2016-04-03'` is not picking up time records from 2016-04-03. If I change it to 2016-04-04, all is well. It seems like it is only the upper limit of the BETWEEN that is not acting like it is inclusive. I have to change the code to use 201-04-04 just to get payroll to process, but will get to the bottom of the issue and correct it later. I am wondering if it is a time zone issue.

**Comment by mkwiatkowski** [comment]
According to docs at http://www.postgresql.org/docs/9.1/static/functions-comparison.html 

```
a BETWEEN x AND y

is equivalent to

a >= x AND a <= y
```

Also, you've set up `clockin` as a "datetime" type column, so '2016-04-03' will be converted to '2016-04-03 00:00:00'. So the only record you can get that matches and is on April 3rd is a record created exactly at midnight (which is very unlikely). You can either add one day to the end date during comparisons like that or force the use of just the date part of the column:

```
date(time_records.clockin) BETWEEN '2016-03-21' AND '2016-04-03'
```

**Comment by Mei152** [comment]
Yes, I had narrowed it down to the data type, but had yet to determine how best to address it. Your explanation is very helpful in that it explains exactly what the mechanics of what is going on are. I made a bit of a mess changing column types on the db and am trying to set things aright.

**Comment by Mei152** [comment]
Wow. Evidently, once you change a column type from a timestamp, with or without timezone, to another date/time type, there is no going back using the GUI, even if the table is empty. Finally had to run a script.

**Comment by mkwiatkowski** [comment]
I already showed you the solution - modifying the query to use only the date part of the datetime column.

**Comment by Mei152** [comment]
Sorry, my last statement should have been put in better context. I was trying to repair damage I did prior to your suggesting a solution. In my experimentation, I changed the clockin column's data type several times, which caused data in that column to be lost. So, I deleted all records, reset the uploaded files' persisted status to false, and tried to re-process everything. The problem was then that persisting time records kept failing. After, a few hours of diagnosing, I discovered that clockin has to be of type timestamp, because it, coupled with some other attributes, is used in a unique constraint. We need the time part of timestamp to give enough spectrum that an employee can clock in more than once on the same day, and still be perceived  by the db as being a unique record for that employee. Learned a lot.

**Comment by Mei152** [comment]
What made the diagnosing so difficult is that the error shown in the flash said record invalid, empnum already exists, which put me on the path of thinking it was a pos_employee persistence problem, rather than a time record persistence problem. This reinforced to me the importance of customized error messages that provide more definition of the issue.
